### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by gammazero**
Refactor the indexer core to be constructed with two stores: resultCache and valueStore. The resultCache is optional, the valueStore is not. The cache is only used to store values that are results of queries and should be of limited size. The valueStore stores all entries regardless of whether they are queried or not.

Refactored common test and benchmark code so that is in its own package that is only imported by tests.  This allows it to be compatible with `go test` and not be built as part of the main library.  Any new store that implements `store.Interface` will be able to use the same tests and benchmarks.

### CODE REVIEW

File: store/pogreb/pogreb.go [hunk 1]
```diff
@@ -163,7 +164,7 @@ func (s *pStorage) RemoveProvider(providerID peer.ID) error {
 // DuplicateEntry checks if the entry already exists in the index. An entry
 // for the same provider but different metadata is not considered
 // a duplicate entry,
-func duplicateEntry(in store.IndexEntry, old []store.IndexEntry) bool {
+func duplicateEntry(in entry.Value, old []entry.Value) bool {
```
Comment (adlrocha) [inline store/pogreb/pogreb.go:167]: We can probably replace this `duplicateEntry` for `entry.ValueInSlice`. They do the same.

File: store/storethehash/storethehash.go [hunk 2]
```diff
@@ -181,7 +182,7 @@ func (s *sthStorage) RemoveProvider(providerID peer.ID) error {
 // DuplicateEntry checks if the entry already exists in the index. An entry
 // for the same provider but different metadata is not considered
 // a duplicate entry.
-func duplicateEntry(in store.IndexEntry, old []store.IndexEntry) bool {
+func duplicateEntry(in entry.Value, old []entry.Value) bool {
```
Comment (adlrocha) [inline store/storethehash/storethehash.go:185]: Same here. We can replace `duplicateEntry` for `entry.ValueInSlice` and remove this.

**Review by adlrocha — APPROVED** [review]

**Comment by adlrocha** [comment]
This looks great, @gammazero. Once we land the refactor in `storetheindex` I think we are good to go. Thanks!

File: store/pogreb/pogreb.go (hunk 1 above)
Comment (gammazero) [reply store/pogreb/pogreb.go:167]: `ValueInSlice` takes pointer arguments, which means it should probably be moved back into the cache implementation, and we should keep `duplicateEntry` here.

File: store/storethehash/storethehash.go (hunk 2 above)
Comment (gammazero) [reply store/storethehash/storethehash.go:185]: same comment as above

File: store/pogreb/pogreb.go (hunk 1 above)
Comment (adlrocha) [reply store/pogreb/pogreb.go:167]: Oh, I see the difference. Then, maybe let's leave it as-is. Up to you. Thanks!

**Pull request merged**